docs/designers-developers/developers/block-api/block-context.md

```diff
@@ -0,0 +1,70 @@
+# Block Context
+
+Block context is a feature which enables ancestor blocks to provide values which can be consumed by descendent blocks within its own hierarchy. Those descendent blocks can inherit this value without resorting to hard-coded values, or without an explicit awareness of the block which provides those values.
+
+This is especially useful in full-site editing where, for example, the contents of a block may depend on the context of the post in which it is relevant. A blogroll template may show excerpts of many different posts. Using block context, there can still be one single "Post Excerpt" block which displays the contents of the post based on an inherited post ID.
+
+If you are familiar with [React Context](https://reactjs.org/docs/context.html), the idea is very similar. In fact, the client-side block editor implementation of block context is a very simple application of React Context. Block context is also supported in server-side `render_callback` implementations, demonstrated in the examples below.
+
+## Defining Block Context
+
+Block context is defined in the registered settings of a block. A block can provide a context value, or consume a value it seeks to inherit.
+
+### Providing Block Context
+
+A block can provide a context value by assigning a `providesContext` property in its registered settings. This should be assigned as an array of the attribute names for attribute values to assign as context. Currently, block context only supports values derived from the block's own attributes. This could be enhanced in the future to support additional sources of context values.
+
+_block.json_
+
+```json
+{
+	"name": "core/post",
+	"attributes": {
+		"postId": {
+			"type": "number"
+		}
+	},
+	"providesContext": [ "postId" ]
```

Comment:
I feel the name here won't always map to attribute names, can we pass an object to change the context name?

Reply:
> I feel the name here won't always map to attribute names, can we pass an object to change the context name?

It was mentioned in passing as a "future enhancement" idea in the original post, but I expanded on a very similar proposal in my comment at https://github.com/WordPress/gutenberg/issues/19685#issuecomment-603848589 .

I chose to keep things simple for this first pass, though I agree it's quite likely it will need to support remapping, especially considering we already are at risk to have conflicts of naming given the lack of namespacing.

I could include an implementation here, or I could create a separate tracking issue as a follow-up task.

Reply:
I'm fine with a follow-up. I do think namespacing is a good idea though, regardless, the reason being:

 - A third-party block adds a new context and name it "comment"
 - Later, Core decides it needs this context

What are your thoughts on these situations?

Reply:
It's hard to know how realistically these specific hypotheticals are to occur.

But I grant that it can be a problem. And it's made especially worse by the fact that currently, there is not even an _option_ to provide a namespace if one opted to choose to, other than to embed something like a prefix in the attribute name itself.

I'm wondering about a slightly altered approach:

- No changes to `providesContext`
- The "context" inherits a namespace by the namespace of block from which it was provided
- As with block registration, `core/` is implied by absence of a namespace

Following your example:

```json
{ "context": [ "comment", "my-plugin/comment" ] }
```

There's a few advantages here:

- The namespace is applied automatically, without intervention of the block implementer.
- Don't _need_ to have the option to "map" custom context names (because of the prior point).
- Core context is still succinct.

The main downside I consider is that there is some element of _magic_ in all of this, specifically in how the namespace would be inherited automatically from the block name.

Reply:
My main downside with this is that a third-party author can't easily say, I'm going to provide an alternative "core/post" context provider

Reply:
> My main downside with this is that a third-party author can't easily say, I'm going to provide an alternative "core/post" context provider in my own block that select posts very specifically.

True. If we _did_ have the ability to map custom context names, they could still do that. And still have this simplified form as the default behavior.

Reply:
I'd be fine with that yes 👍

Reply:
As I'm working through an implementation, I'm encountering another question related to this: If a context key is namespaced, how should it be referenced in `edit` or `render_callback` implementations of a block? Should it use the fully-qualified (namespaced) key? Or the name as it was provided by the ancestor?

```js
function edit( { context } ) {
    const a = context.comment;
    const b = context[ 'my-plugin/comment' ];
}
```

There are a couple considerations:

- There may be occasions where a consumer wants to use a value from multiple namespaces, so retaining the namespace would be necessary to disambiguate.
- Namespaced form is not quite as nice to use, both in verboseness and that in JavaScript, one must use square-bracket syntax for property access.
- It makes most sense to consider omitting the namespace in considering that the provider doesn't explicitly add it, and for its purposes it treats it the same as an attribute name. 
   - `{ "name": "my-plugin/comment", "providesContext": [ "comment" ] }`
- The implementation of repeatedly adding and removing a namespace is cumbersome, which may speak to a broader problem of this being complex (and therefore risking to be not well-understood).

My personal thoughts:

- This is only an issue if we actually have namespacing. We could still choose not to.
- That said, I've come around to the idea and think that it is "safer" to anticipate potential conflicts and address them through namespacing.
- While it is not quite as expressive, I think referencing the fully-qualified name is probably the simplest both in implementation and in removing some non-obvious magic by limiting how often these values are transformed.

I do worry about how the ancestor would still define the context it provides without a namespace, and how that namespace would be automatically injected after-the-fact. I might wonder if it would be simplest to just remove this array form, and _only_ support a format where a provider defines a mapping of context key to attribute name.

Something like:

```json
{
	"name": "my-plugin/record",
	"attributes": {
		"recordId": {
			"type": "number"
		}
	},
	"providesContext": {
		"my-plugin/recordId": "recordId"
	}
}
```

I think this works well in theory. In practice, I could worry that if we didn't enforce the namespacing, plugins may not choose to include it. It might be an acceptable risk.

One particularly nice aspect of this approach is that it effectively removes all magic, and "namespace" doesn't need to be a formal concept of the context arrangement. It's merely a convention.

Reply:
> As I'm working through an implementation, I'm encountering another question related to this: If a context key is namespaced, how should it be referenced in `edit` or `render_callback` implementations of a block? Should it use the fully-qualified (namespaced) key? Or the name as it was provided by the ancestor?

I shared my comment above after I'd largely finished the implementation of this automatic namespacing. You can see it in the history at 151d425fe, which I proceeded to revert in c78e6eea9.

Instead, what I landed at was 3950a8555.

Note:

- Context is always provided as a mapping.
- There's nothing about this which requires a namespace.
   - This keeps the implementation and referencing simple. There is no magic, no transforming of context names to or from formats including or not including the namespace.
   - To address the worry that plugins "won't do it", the documentation was changed to use examples from the perspective of a plugin's block implementation, granting an opportunity to clarify these conventions as the default approach.
   - Core usage can still omit the namespace (because, again, the namespace is merely convention). Because it's now an object mapping, plugins can easily override the core values if they so choose.

Reply:
> Context is always provided as a mapping.

I like the reasoning and I also agree that is probably the safest and the simplest solution. I do wonder about core omitting the namespace though and whether we should add a regex check like we do for core blocks.

Reply:
> > Context is always provided as a mapping.
> 
> I like the reasoning and I also agree that is probably the safest and the simplest solution. I do wonder about core omitting the namespace though and whether we should add a regex check like we do for core blocks.

Oh, I had thoughts on this, which I neglected to include.

If we wanted to, we could have some restrictions in place, and refuse / warn to accept a context key from `providesContext` if it doesn't match a pattern we expect.